### PR TITLE
Provide optional automatic ledger/wallet backups before an upgrade

### PR DESCRIPTION
**Pull request by SergiySW**
https://github.com/nanocurrency/nano-node/issues/1690

### CODE REVIEW

File: nano/node/lmdb/lmdb.cpp [hunk 1]
```diff
@@ -60,9 +60,15 @@ txn_tracking_enabled (txn_tracking_config_a.enable)
...
 		// open inactive nodes which can otherwise be locked here if there is a long write
-		// (can be a few minutes with the --fastbootstrap flag for instance)
+		// (can be a few minutes with the --fast_bootstrap flag for instance)
 		if (!is_fully_upgraded)
 		{
+			// Backup database before upgrade
+			if (backup_before_upgrade)
+			{
+				auto backup_path = path_a.parent_path () / "backup.upgrade.ldb";
+				mdb_env_copy (env, backup_path.string ().c_str ());
```
Comment (wezrule) [inline nano/node/lmdb/lmdb.cpp]: If they wanted a backup we should probably inform them and exit the node if the backup fails (lack of disk space etc), so they can rectify it and try again.

**Comment by SergiySW** [comment]
@guilhermelawless imagine sevweral data.ldb backups size

**Comment by guilhermelawless** [comment]
That's a good point. Can we then follow some other convention, for example `data.ldb.bak` and `wallets.ldb.bak`?

**Comment by SergiySW** [comment]
I guess we can use config_backup_15571460497935531.json style for wallets (usually small size) & other for data.ldb

File: nano/node/lmdb/lmdb.cpp (hunk 1 above)
Comment (SergiySW) [reply nano/node/lmdb/lmdb.cpp]: Good point

**Comment by schenkty** [comment]
Why wouldn't this happen by default? Many of us service operators have run into countless issues where something is corrupt after an upgrade and we have to start from scratch. Would be awesome if it could backup automatically then I suppose delete the backup once everything is successful and just make it optional to retain the backup post upgrade

**Comment by guilhermelawless** [comment]
@schenkty All it takes is changing a config flag then you're set for every future upgrade

**Comment by cryptocode** [comment]
Adding toml label as `backup_before_upgrade` upgrade change will need to be reverted and placed into toml base file.

**Comment by guilhermelawless** [comment]
This will have to be changed as #2203 was merged which creates the new config upgrade path.

**Pull request merged**